# Homework 1 - data validation & cleaning (deadline 30. 10. 2022, 23:59)

In short, the main task is to clean The Metropolitan Museum of Art Open Access dataset.
  
> The instructions are not given in details: It is up to you to come up with ideas on how to fulfill the particular tasks as best as you can!

However, we **strongly recommend and require** the following:
* Follow the assignment step by step. Number each step.
* Most steps contain the number of features that should be treated. You can preprocess more features, however, it does not mean that the teacher will give you more points. Focus on quality, not quantity.
* Properly comment all your steps. Comments are evaluated for 2 points of the total together with the final presentation of the solution. However, it is not desirable to write novels! 
* This task is the most challenging of the three, both timewise and computationally. Do not leave it to the last minute.
* Hand in a notebook that has already been run (i.e. do not delete outputs before handing in).

## What are you supposed to do:

  1. Download the dataset MetObjects.csv from the repository https://github.com/metmuseum/openaccess/.
  2. Check consistency (i.e. that same things are represented in the same way) of at least **three features** where you expect problems (include "Object Name" feature). You can propose how to clean the selected features, however do not apply cleaning (in your interest) 🙂 _(1.5 points)_
  3. Select at least **two features** where you expect integrity problems (describe your choice) and check integrity of those features. By integrity we mean correct logical relations between features (e.g. female names for females only). _(2 points)_
  4. Convert at least **five features** to a proper data type. Choose at least one numeric, one categorical and one datetime. _(1.5 points)_
  5. Find some outliers and describe your method. _(3 points, depends on creativity)_
  6. Detect missing data in at least **three features**, convert them to a proper representation (if they are already not), and impute missing values in at least **one feature**. _(1 + 3 points, depends on creativity)_
  7. Focus more precisely on the cleaning of the "Medium" feature. As if you were to use it in KNN algorithm later. _(2 points)_
  8. Focus on the extraction of physical dimensions of each item (width, depth and height in centimeters) from the "Dimensions" feature. _(2 points)_

All your steps and following code **have to be commented!** Comments are evaluated for 2 points together with the final presentation of the solution.

**If you do all this properly, you will obtain 18 points.**

## Comments

  * Please follow the technical instructions from https://courses.fit.cvut.cz/NI-PDD/homeworks/index.html.
  * If the reviewing teacher is not satisfied, she can (!) give you another chance to rework your homework and to obtain more points. However, this is not a given, so do your best! :)

In [1]:
# importing basic libraries
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

# 1. Download the dataset MetObjects.csv from the repository https://github.com/metmuseum/openaccess/.

In [2]:
# importing data
df = pd.read_csv('MetObjects.csv', sep=',') # load data as pandas DataFrame

/var/folders/gm/6zjcwbd164d1v4gmx73009pr0000gn/T/ipykernel_84845/4046042708.py:2: DtypeWarning: Columns (5,7,10,11,12,13,14,34,35,36,37,38,39,40,41,42,43,44,45,46) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('MetObjects.csv', sep=',') # load data as pandas DataFrame


## Random data exploration

In [3]:
df.shape

(477804, 54)

In [90]:
df

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
1            1979.486.1         False             False             False   
2            1980.264.5         False             False             False   
3              67.265.9         False             False             False   
4             67.265.10         False             False             False   
5             67.265.11         False             False             False   
...                 ...           ...               ...               ...   
860869      23.112.2893         False             False              True   
860870      23.112.2894         False             False              True   
860871      53.600.1434         False             False             False   
860872      23.112.2895         False             False              True   
860873      23.112.2896         False             False              True   

          Gallery Number           Department                 AccessionYear  \
Object ID                                                                     
1                    NaN    The American Wing 1970-01-01 00:00:00.000001979   
2                    NaN    The American Wing 1970-01-01 00:00:00.000001980   
3                    NaN    The American Wing 1970-01-01 00:00:00.000001967   
4                    NaN    The American Wing 1970-01-01 00:00:00.000001967   
5                    NaN    The American Wing 1970-01-01 00:00:00.000001967   
...                  ...                  ...                           ...   
860869               NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860870               NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860871               NaN  Drawings and Prints 1953-01-01 00:00:00.000000000   
860872               NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860873               NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   

          Object Name                                              Title  \
Object ID                                                                  
1                Coin                       One-dollar Liberty Head Coin   
2                Coin                       Ten-dollar Liberty Head Coin   
3                Coin                         Two-and-a-Half Dollar Coin   
4                Coin                         Two-and-a-Half Dollar Coin   
5                Coin                         Two-and-a-Half Dollar Coin   
...               ...                                                ...   
860869        Drawing                                     Phaeton #24567   
860870        Drawing             Stanhope Phaeton #25538-25539 (#21222)   
860871          Print  Forest landscape with cattle drinking, a woman...   
860872        Drawing                           Phaeton with folding top   
860873        Drawing                                   Stanhope Phaeton   

          Culture  ... Rights and Reproduction  \
Object ID          ...                           
1             NaN  ...                     NaN   
2             NaN  ...                     NaN   
3             NaN  ...                     NaN   
4             NaN  ...                     NaN   
5             NaN  ...                     NaN   
...           ...  ...                     ...   
860869        NaN  ...                     NaN   
860870        NaN  ...                     NaN   
860871        NaN  ...                     NaN   
860872        NaN  ...                     NaN   
860873        NaN  ...                     NaN   

                                               Link Resource  \
Object ID                                                      
1           http://www.metmuseum.org/art/collection/search/1   
2           http://www.metmuseum.org/art/collection/search/2   
3           http://www.metmuseum.org/art/collection/search/3   
4           http://w

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 477804 entries, 1 to 860873
Data columns (total 55 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Object Number            477804 non-null  object        
 1   Is Highlight             477804 non-null  bool          
 2   Is Timeline Work         477804 non-null  bool          
 3   Is Public Domain         477804 non-null  bool          
 4   Gallery Number           51776 non-null   object        
 5   Department               477804 non-null  category      
 6   AccessionYear            474248 non-null  datetime64[ns]
 7   Object Name              476113 non-null  object        
 8   Title                    448619 non-null  object        
 9   Culture                  207379 non-null  object        
 10  Period                   90956 non-null   object        
 11  Dynasty                  23233 non-null   object        
 12  Reign           

In [92]:
df.nunique()

Object Number              474872
Is Highlight                    2
Is Timeline Work                2
Is Public Domain                2
Gallery Number                609
Department                     19
AccessionYear                 280
Object Name                 28449
Title                      239091
Culture                      7180
Period                       1873
Dynasty                       403
Reign                         389
Portfolio                    3563
Constituent ID              46224
Artist Role                  6837
Artist Prefix                8046
Artist Display Name         64531
Artist Display Bio          52439
Artist Suffix                2729
Artist Alpha Sort           64489
Artist Nationality           5037
Artist Begin Date           29716
Artist End Date             29685
Artist Gender                 284
Artist ULAN URL             37235
Artist Wikidata URL         39362
Object Date                 32697
Object Begin Date            2073
Object End Dat

In [7]:
df.describe(include='all')

Object Number Is Highlight Is Timeline Work Is Public Domain  \
count         477804       477804           477804           477804   
unique        474872            2                2                2   
top           62.635        False            False             True   
freq               4       475320           469816           245307   
mean             NaN          NaN              NaN              NaN   
std              NaN          NaN              NaN              NaN   
min              NaN          NaN              NaN              NaN   
25%              NaN          NaN              NaN              NaN   
50%              NaN          NaN              NaN              NaN   
75%              NaN          NaN              NaN              NaN   
max              NaN          NaN              NaN              NaN   

            Object ID Gallery Number           Department  AccessionYear  \
count   477804.000000          51776               477804       474248.0   
unique            NaN            609                   19          280.0   
top               NaN            774  Drawings and Prints         1963.0   
freq              NaN           7054               167152        34280.0   
mean    387582.167811            NaN                  NaN            NaN   
std     237374.736434            NaN                  NaN            NaN   
min          1.000000            NaN                  NaN            NaN   
25%     210119.750000            NaN                  NaN            NaN   
50%     371186.500000            NaN                  NaN            NaN   
75%     563883.250000            NaN                  NaN            NaN   
max     860873.000000            NaN                  NaN            NaN   

       Object Name                                          Title  ...  \
count       476113                                         448619  ...   
unique       28449                                         239091  ...   
top          Print  Terracotta fragment of a kylix (drinking cup)  ...   
freq         99439                                           6415  ...   
mean           NaN                                            NaN  ...   
std            NaN                                            NaN  ...   
min            NaN                                            NaN  ...   
25%            NaN                                            NaN  ...   
50%            NaN                                            NaN  ...   
75%            NaN                                            NaN  ...   
max            NaN                                            NaN  ...   

                    River Classification  \
count                2095         399598   
unique                229           1212   
top     Upper Sepik River         Prints   
freq                  362          80981   
mean                  NaN            NaN   
std                   NaN            NaN   
min                   NaN            NaN   
25%                   NaN            NaN   
50%                   NaN            NaN   
75%                   NaN            NaN   
max                   NaN            NaN   

                                  Rights and Reproduction  \
count                                               24198   
unique                                               1415   
top     © Walker Evans Archive, The Metropolitan Museu...   
freq                                                 7314   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                           Link Resource  \
count                         

In [8]:
df.duplicated().sum()

0

In [9]:
# df.rename(columns={'Unnamed: 0': 'id'}, inplace=True)

In [10]:
df.set_index('Object ID', inplace=True) # use IDs as index
df.duplicated().sum()

0

No duplicates

#  2. Check consistency (i.e. that same things are represented in the same way) of at least **three features** where you expect problems (include "Object Name" feature). You can propose how to clean the selected features, however do not apply cleaning (in your interest) 🙂 _(1.5 points)_

## 2a Since the assigment says to include the `Object Name`, lets start with that

In [11]:
df['Object Name'].describe()

count     476113
unique     28449
top        Print
freq       99439
Name: Object Name, dtype: object

In [12]:
# df['Title_len'] = df.apply(lambda x: len(x['Title']), axis=1)
#df['Object_name_len'] = df['Object Name'].apply(lambda x: len(str(x)))
df['Object_name_len'] = df['Object Name'].str.len()

In [13]:
df['Object_name_len'].describe()

count    476113.000000
mean         10.476992
std           8.560064
min           2.000000
25%           5.000000
50%           7.000000
75%          13.000000
max          80.000000
Name: Object_name_len, dtype: float64

In [14]:
# df.groupby('Object Name').nunique().sort_values('Object_name_len')
# df.loc[df.groupby('Object Name').Object_name_len.nsmallest(3).index]
# dataframe with records without duplicated 'Object name'
df_dedup_name = df.drop_duplicates(subset=['Object Name'])

In [15]:
# get 20 of the shortest Object names
df_dedup_name.loc[df_dedup_name.Object_name_len.nsmallest(20).index]['Object Name']

Object ID
24857      Ax
248056     As
500526     Ūd
500580     Po
500582     Lo
500589     Bo
500606     Gu
501092     Ty
503211     So
503339     Em
503765     Se
504702     Yu
327       Mug
329       Keg
688       Cup
896       Box
948       Jar
1401      Cap
2080      Pin
2126      Urn
Name: Object Name, dtype: object

after looking these up on the internet, the names do seem to make sense

altough, wait, 'Ax'? I know it as 'Axe', I wonder if that is there as well?

In [16]:
df.loc[df['Object Name'] == 'Ax']

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
24857            69.156         False             False              True   
25577          04.3.467         False              True              True   
32944          23.276.2         False             False              True   
37759       2000.284.58         False             False              True   
38627       2000.284.61         False             False              True   
...                 ...           ...               ...               ...   
555849        07.231.14         False             False             False   
557942          17.6.65         False             False             False   
569247          26.2.65         False             False             False   
571171      09.180.1413         False             False             False   
571857          29.2.11         False             False             False   

          Gallery Number      Department AccessionYear Object Name  \
Object ID                                                            
24857              379.0  Arms and Armor        1969.0          Ax   
25577                NaN  Arms and Armor        1904.0          Ax   
32944              370.0  Arms and Armor        1923.0          Ax   
37759              244.0       Asian Art        2000.0          Ax   
38627              244.0       Asian Art        2000.0          Ax   
...                  ...             ...           ...         ...   
555849             109.0    Egyptian Art        1907.0          Ax   
557942               NaN    Egyptian Art        1917.0          Ax   
569247               NaN    Egyptian Art        1926.0          Ax   
571171               NaN    Egyptian Art        1909.0          Ax   
571857               NaN    Egyptian Art        1929.0          Ax   

                             Title    Culture  ...   Classification  \
Object ID                                      ...                    
24857                           Ax     Syrian  ...  Shafted Weapons   
25577                   Dervish Ax    Turkish  ...  Shafted Weapons   
32944      Ax of the Palstave Type    British  ...    Miscellaneous   
37759                          NaN    Vietnam  ...        Metalwork   
38627                          NaN  Indonesia  ...        Sculpture   
...                            ...        ...  ...              ...   
555849                      Ax (?)        NaN  ...              NaN   
557942                          Ax        NaN  ...              NaN   
569247                          Ax        NaN  ...              NaN   
571171                          Ax        NaN  ...              NaN   
571857                          Ax        NaN  ...              NaN   

          Rights and Reproduction  \
Object ID                           
24857                         NaN   
25577                         NaN   
32944                         NaN   
37759                         NaN   
38627                         NaN   
...                           ...   
555849                        NaN   
557942                        NaN   
569247                        NaN   
571171                        NaN   
571857                        NaN   

                                               Link Resource  \
Object ID                                                      
24857      http://www.metmuseum.org/art/collection/search...   
25577      http://www.metmuseum.org/art/collection/search...   
32944      http://www.metmuseum.org/art/collection/search...   
37759      http://www.metmuseum.org/art/collection/search...   
38627      http://www.metmuseum.org/art/collection/search...   
...                                                      ...   
555849     http://www.metmuseum.org/art/collection/search...   
557942     http://www.metmuseum.org/art/collection/search...   
569247     http://www.metmuseum.org/art/collection/

In [17]:
df.loc[df['Object Name'] == 'Axe']

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
25066           04.3.40         False             False             False   
25578          04.3.470         False             False             False   
27065         14.25.464         False             False             False   
27882           46.79.2         False             False             False   
30784        36.25.1800         False             False             False   
...                 ...           ...               ...               ...   
327471       1992.149.7         False             False              True   
329076           1982.5          True              True              True   
332605      1989.281.45         False             False              True   
561585         20.3.278         False             False              True   
635754       2014.523.3         False             False             False   

          Gallery Number                                 Department  \
Object ID                                                             
25066                NaN                             Arms and Armor   
25578                NaN                             Arms and Armor   
27065                NaN                             Arms and Armor   
27882                NaN                             Arms and Armor   
30784                NaN                             Arms and Armor   
...                  ...                                        ...   
327471               NaN                   Ancient Near Eastern Art   
329076             403.0                   Ancient Near Eastern Art   
332605               NaN                   Ancient Near Eastern Art   
561585             106.0                               Egyptian Art   
635754               NaN  Arts of Africa, Oceania, and the Americas   

          AccessionYear Object Name  \
Object ID                             
25066            1904.0         Axe   
25578            1904.0         Axe   
27065            1914.0         Axe   
27882            1946.0         Axe   
30784            1936.0         Axe   
...                 ...         ...   
327471           1992.0         Axe   
329076           1982.0         Axe   
332605           1989.0         Axe   
561585           1920.0         Axe   
635754           2014.0         Axe   

                                                       Title  \
Object ID                                                      
25066                                                    Axe   
25578                                                    Axe   
27065                                                    Axe   
27882                                                    Axe   
30784                                                    Axe   
...                                                      ...   
327471                                   Axe with shaft hole   
329076     Shaft-hole axe head with bird-headed demon, bo...   
332605                              Fragmentary axe head (?)   
561585                                                   Axe   
635754                                        Ceremonial Axe   

                                           Culture  ...  \
Object ID                                           ...   
25066                                       German  ...   
25578                                       German  ...   
27065                                      Swedish  ...   
27882                                    Norwegian  ...   
30784                          Indian, south India  ...   
...                                            ...  ...   
327471     Bactria-Margiana Archaeological Complex  ...   
329076     Bactria-Margiana Archaeological Complex  ...   
332605     Bactria-Margiana Archaeological Complex  ...   
561585                                         NaN  ...   
635754                             

Both of these names are indeed used for the same thing. (By checking the department "Arms and Armor" we can make sure, that those 2 different names, do indeed mean the same thing)
Solving this problem might be a bit difficult, but maybe we could compare against a list of synonyms or translations

In [18]:
# get 20 of the shortest Object names
df_dedup_name.loc[df_dedup_name.Object_name_len.nlargest(20).index]['Object Name']
# Hmm, kinda SUS

Object ID
24684     Double-barreled flintlock shotgun with exchang...
24977     Nineteen ornamental fittings from a horse harn...
30379     Ammunition belt (Kamr) Object Cards.  S. martz...
30958     Blade and for a hatchet (Nata) and mounting fo...
34765     Set of Pommel (Fuchigashira), Sword grip ornam...
35744     Top two lames from a right thigh defense (Cuis...
545279    Pall, sheet, Wah, rather coarse, light weight,...
545619    String of 16 spherical, 1 barrel and 1 cylindr...
552577    Cone, round, Padiamunnebnesuttawy, chancellor,...
553399    Scarab, Menkheperra, nsw bit, royal title, mry...
554965    Cone, Ptahmose, circular impression, high prie...
555025    Cone, rectangular impression, Neferhotep, scri...
555048    Cone, circular impression, Kaemamun, prophet, ...
555113    Cylinder seal with name of Amenemhat II and th...
556167    Cylinder seal with cartouches of Nubkaure (Ame...
557109    Scarab, pseudo-inscription in oval, hieroglyph...
557815    Facsimile, Panehsy (

From example above we can see the longest objects names.

Lets take a look at `Scarab, Menkheperra, nsw bit, royal title, mry imn, beloved of Amun, Thutmose II` and see if there is only `Scarab`

In [19]:
df.loc[df['Object Name'] == 'Scarab']

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
243605       74.51.4193         False             False              True   
243606       74.51.4194         False             False              True   
244873           81.6.2         False             False             False   
244875           81.6.4         False             False              True   
244876           81.6.5         False             False             False   
...                 ...           ...               ...               ...   
575445         23.3.183         False             False              True   
575446         23.3.187         False             False              True   
575485         23.3.260         False             False              True   
575486         23.3.275         False             False              True   
575487         23.3.283         False             False              True   

          Gallery Number           Department AccessionYear Object Name  \
Object ID                                                                 
243605             171.0  Greek and Roman Art        1874.0      Scarab   
243606             171.0  Greek and Roman Art        1874.0      Scarab   
244873               NaN  Greek and Roman Art        1881.0      Scarab   
244875             170.0  Greek and Roman Art        1881.0      Scarab   
244876               NaN  Greek and Roman Art        1881.0      Scarab   
...                  ...                  ...           ...         ...   
575445             117.0         Egyptian Art        1923.0      Scarab   
575446             117.0         Egyptian Art        1923.0      Scarab   
575485             117.0         Egyptian Art        1923.0      Scarab   
575486             117.0         Egyptian Art        1923.0      Scarab   
575487             117.0         Egyptian Art        1923.0      Scarab   

                         Title            Culture  ...        Classification  \
Object ID                                          ...                         
243605            Agate scarab              Greek  ...                  Gems   
243606                  Scarab            Cypriot  ...  Gems|Gold and Silver   
244873                  Scarab              Roman  ...                  Gems   
244875     Banded agate scarab           Etruscan  ...                  Gems   
244876                  Scarab  Greek or Etruscan  ...                  Gems   
...                        ...                ...  ...                   ...   
575445                  Scarab                NaN  ...                   NaN   
575446                  Scarab                NaN  ...                   NaN   
575485                  Scarab                NaN  ...                   NaN   
575486                  Scarab                NaN  ...                   NaN   
575487                  Scarab                NaN  ...                   NaN   

          Rights and Reproduction  \
Object ID                           
243605                        NaN   
243606                        NaN   
244873                        NaN   
244875                        NaN   
244876                        NaN   
...                           ...   
575445                        NaN   
575446                        NaN   
575485                        NaN   
575486                        NaN   
575487                        NaN   

                                               Link Resource  \
Object ID                                                      
243605     http://www.metmuseum.org/art/collection/search...   
243606     http://www.metmuseum.org/art/collection/search...   
244873     http://www.metmuseum.org/art/collection/search...   
244875     http://www.metmuseum.org/art/collection/search...   
244876     http://www.metmuseum.org/art/collection/search...   
...                                                      ... 

Aha, there is also only `Scarab`, therefore another consistency problem.
To remove this kind of incostistency problem we could use `.apply(lambda x: x.split(',')[0])`, same thing should be checked and done for other special characters (`|`, `.`, `/`, ... )

Check for unprintable characters (only LF newline in this case)

In [20]:
from string import printable
# df.loc[str(df['Object Name']).isalnum()]
# df.loc[df['Object Name'].apply(lambda x: not str(x).isalnum())]
# df.loc[df['Object Name'].apply(lambda x: set(str(x)).difference(printable))]
df.loc[df['Object Name'].apply(lambda x: '\n' in str(x))]


Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
3580             16.177         False             False              True   
3658       1971.180.65b         False             False             False   
5612         1975.400.3         False             False              True   
12890          46.67.84         False             False              True   
19200           2007.43         False             False              True   
...                 ...           ...               ...               ...   
788902        21.91.243         False             False              True   
788903        21.91.244         False             False              True   
820446     26.50.1(146)         False             False             False   
822562         28.99.29         False             False              True   
845522     2020.373a, b         False             False              True   

          Gallery Number           Department AccessionYear  \
Object ID                                                     
3580                 751    The American Wing        1916.0   
3658                 741    The American Wing        1971.0   
5612                 704    The American Wing        1975.0   
12890                730    The American Wing        1946.0   
19200                NaN    The American Wing        2007.0   
...                  ...                  ...           ...   
788902               NaN  Drawings and Prints        1921.0   
788903               NaN  Drawings and Prints        1921.0   
820446               NaN  Drawings and Prints        1926.0   
822562               NaN       Arms and Armor        1928.0   
845522               NaN          Islamic Art        2020.0   

                                                 Object Name  \
Object ID                                                      
3580       Exterior Doorway from One West Broad Street,\r...   
3658                                          Fire Tongs\r\n   
5612                                               Plate\r\n   
12890      Painting\r\nFrigate Presiden\r\nFrigate Presid...   
19200        Panel\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\nWindow   
...                                                      ...   
788902                                             Print\r\n   
788903                                             Print\r\n   
820446                                     Drawing \r\nPrint   
822562                                      Crossbow nut\r\n   
845522                            Mirror case with pouch\r\n   

                                                       Title  \
Object ID                                                      
3580       Exterior Doorway from One West Broad Street, S...   
3658                                              Fire Tongs   
5612                                                   Plate   
12890      Fight Between the U.S. Frigate President and H...   
19200                                                 Spring   
...                                                      ...   
788902                                  Chinoiserie Ornament   
788903                                  Chinoiserie Ornament   
820446                       Five dancing or fighting cupids   
822562                                          Crossbow Nut   
845522     Mirror Case Depicting the Meeting of Nasir al-...   

                                    Culture  ...  \
Object ID                                    ...   
3580                               American  ...   
3658                                    NaN  ...   
5612                               American  ...   
12890                              American  ...   
19200                              American  ...   
...                                     ...  ...   
788902                                  NaN  ...   
788903                                  NaN  ...   
82

We can see a name such as `Bowl\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n`
So we should clean the names of any unprintable characters. For this particular case a simple `name.strip()` would work, but a better solution is needed for all rows

In [21]:
# df_dedup_name['Object_name_commas_count'] = df_dedup_name['Object Name'].apply(lambda x: str(x).count(','))
# df_dedup_name.Object_name_commas_count.describe()

Lets take a look if case sensitivity plays a role

In [22]:
# get number of duplicate names
df.duplicated(subset=['Object Name']).sum()

449354

In [23]:
# get number of duplicated names when case sensitivity does not matter
df['Object Name'].apply(lambda x: str(x).upper()).duplicated().sum()

450202

We can see that there are inconsistencies in case sensitivity. We should converted all rows in some way (all upper, all lower, 1st upper, ...)

## 2b

Lets decide on 2nd column. When I look back at the initial data exploration, 1st thing that seems very suspicious is `Artist Gender,284` when looking at number of unique values, so `Artist Gender` it is.

In [24]:
# get list of unique values
df_dedup_artist_gender = df.drop_duplicates(subset=['Artist Gender'])
df_dedup_artist_gender

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                    
1             1979.486.1         False             False             False   
35               04.1a–c          True              True             False   
282            25.173a–o          True              True             False   
311              1984.98         False             False             False   
444            18.110.64         False              True              True   
...                  ...           ...               ...               ...   
838183       60.502.7(4)         False             False             False   
840803        40.101(56)         False             False             False   
849227      62.600.15(2)         False             False             False   
853788         66.540.34         False             False             False   
858033     1991.1073.232         False             False             False   

          Gallery Number           Department AccessionYear Object Name  \
Object ID                                                                 
1                    NaN    The American Wing        1979.0        Coin   
35                   706    The American Wing        1904.0        Vase   
282                  700    The American Wing        1925.0      Window   
311                  701    The American Wing        1984.0    Baluster   
444                  729    The American Wing        1918.0    Bedstead   
...                  ...                  ...           ...         ...   
838183               NaN  Drawings and Prints        1960.0        Book   
840803               NaN  Drawings and Prints        1940.0       Print   
849227               NaN  Drawings and Prints        1962.0        Book   
853788               NaN  Drawings and Prints          1966        Book   
858033               NaN  Drawings and Prints          1991        Book   

                                                       Title   Culture  ...  \
Object ID                                                               ...   
1                               One-dollar Liberty Head Coin       NaN  ...   
35                                            The Adams Vase  American  ...   
282                                         Autumn Landscape  American  ...   
311        Baluster from the Guaranty Building, Buffalo, ...  American  ...   
444                                                 Bedstead  American  ...   
...                                                      ...       ...  ...   
838183                           Illustrated London Almanack       NaN  ...   
840803               Artist Color Proof Associates (leaflet)       NaN  ...   
849227       The Life of Her Most Gracious Majesty the Queen       NaN  ...   
853788     A London Garland, Selected from Five Centuries...       NaN  ...   
858033                 Original Etchings by American Artists       NaN  ...   

          Classification Rights and Reproduction  \
Object ID                                          
1                    NaN                     NaN   
35                   NaN                     NaN   
282                  NaN                     NaN   
311                  NaN                     NaN   
444                  NaN                     NaN   
...                  ...                     ...   
838183             Books                     NaN   
840803            Prints                     NaN   
849227             Books                     NaN   
853788             Books                     NaN   
858033             Books                     NaN   

                                               Link Resource  \
Object ID                                                      
1           http://www.metmuseum.org/art/collection/search/1   
35         http://www.metmuseum.org/art/collection/search/35   
282        http://www.metmuseum.org/art/collection/search..

We can see that it is a big mess, with values like `||||||||||||||Female|||||||`
To find all inconsistencies I'll actually try to clean the data

In [25]:
# first convert NaN into empty string
df_dedup_artist_gender = df_dedup_artist_gender.fillna('')

In [26]:
def clean_artist_gender(text: str) -> str:
    # split string into list by `|`
    split_list = text.split('|')
    # remove empty substrings
    split_list = list(filter(len, split_list))
    # deduplicate substrings
    split_list = list(set(split_list))
    # make into string. If there are multiple substrings in list, join it with `, ` (Spoiler there arent any anyway)
    final_value = ", ".join(split_list)
    return final_value

In [27]:
df_dedup_artist_gender = df_dedup_artist_gender['Artist Gender'].apply(lambda x: clean_artist_gender(str(x)))
df_dedup_artist_gender

Object ID
1               
35              
282       Female
311             
444             
           ...  
838183    Female
840803    Female
849227    Female
853788    Female
858033    Female
Name: Artist Gender, Length: 285, dtype: object

In [28]:
# lets dedup again
df_dedup_artist_gender = df_dedup_artist_gender.drop_duplicates()
df_dedup_artist_gender

Object ID
1            
282    Female
Name: Artist Gender, dtype: object

### yikes!
I just noticed that use of the `||||||||||||||Female|||||||` is because there are multiple authors, which makes it not a consistency problem. (It would be better to check if for all author columns the number of `|` is the same in one row, but I am not gonna bother here)
However I'll still count this one as incorrect since there are is no info if the artist is a male. Maybe all null values are male, but we cannot say for sure.

## 2c

I could pick any author related column, but let's look at smth else.
Again, looking at the data exploration, I notice `Object Date`, so lets explore that

In [30]:
# get list of unique values
df_dedup_object_date = df.drop_duplicates(subset=['Object Date'])
df_dedup_object_date

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
1            1979.486.1         False             False             False   
2            1980.264.5         False             False             False   
3              67.265.9         False             False             False   
10           1979.486.3         False             False             False   
11           1979.486.2         False             False             False   
...                 ...           ...               ...               ...   
858211       67.539.162         False             False             False   
858213     1980.1124.25         False             False             False   
858241     17.3.3442-83         False             False             False   
858251       1976.622.1         False             False             False   
860871      53.600.1434         False             False             False   

          Gallery Number           Department AccessionYear     Object Name  \
Object ID                                                                     
1                    NaN    The American Wing        1979.0            Coin   
2                    NaN    The American Wing        1980.0            Coin   
3                    NaN    The American Wing        1967.0            Coin   
10                   NaN    The American Wing        1979.0            Coin   
11                   NaN    The American Wing        1979.0            Coin   
...                  ...                  ...           ...             ...   
858211               NaN  Drawings and Prints          1967           Print   
858213               NaN  Drawings and Prints          1980           Print   
858241               NaN  Drawings and Prints          1917           Print   
858251               NaN  Drawings and Prints          1976  Book, pamphlet   
860871               NaN  Drawings and Prints          1953           Print   

                                                       Title Culture  ...  \
Object ID                                                             ...   
1                               One-dollar Liberty Head Coin     NaN  ...   
2                               Ten-dollar Liberty Head Coin     NaN  ...   
3                                 Two-and-a-Half Dollar Coin     NaN  ...   
10                    Two-and-a-half-dollar Indian Head Coin     NaN  ...   
11                   Two-and-a-half-dollar Liberty Head Coin     NaN  ...   
...                                                      ...     ...  ...   
858211                         Ornamental frame with figures     NaN  ...   
858213                             Portrait of Benjamin West     NaN  ...   
858241            Six guineas entrance and a guinea a lesson     NaN  ...   
858251              Craftsman Catalog of Dog Goods, No. 1-25     NaN  ...   
860871     Forest landscape with cattle drinking, a woman...     NaN  ...   

          Classification Rights and Reproduction  \
Object ID                                          
1                    NaN                     NaN   
2                    NaN                     NaN   
3                    NaN                     NaN   
10                   NaN                     NaN   
11                   NaN                     NaN   
...                  ...                     ...   
858211            Prints                     NaN   
858213            Prints                     NaN   
858241            Prints                     NaN   
858251             Books                     NaN   
860871            Prints                     NaN   

                                               Link Resource  \
Object ID                                                      
1           http://www.metmuseum.org/art/collection/search/1   
2           http://www.metmuseum.org/art/collection/search/2   
3           http://www.metmuseum.org/art/coll

out of some examples there are few ways of representations

```
1853
1909-27
1665-1700
ca. 1887
after 1866
ca. 1902–3
19th century
1821–ca. 1843
1716 or 1719 (edition)
1787 (?)
1879, cast probably 1880
September 25, 1973
```

This makes it clear the ranges and uncertainty are represented in multiple ways => inconsistency

# 3. Select at least **two features** where you expect integrity problems (describe your choice) and check integrity of those features. By integrity we mean correct logical relations between features (e.g. female names for females only). _(2 points)_

## 3A
During step 2c when exploring `Object date` I notice 2 other columns `Object Begin Date` and `Object end Date`, I think that there could be integrity problems between those, since there should be direct relationship between these 3. Lets explore
Lets try to create our own `Object Date` from the other two columns

In [44]:
# convert `Object Date` to string
df['Object Date'] = df['Object Date'].astype(str)

In [55]:
def combine_date(begin: str, end: str) -> str:
    if begin == end:
        return str(begin)
    if begin[0:3] == end[0:3]:
        return f"{begin}–{end[3:]}"
    if begin[0:2] == end[0:2]:
        return f"{begin}–{end[2:]}"
    return f"{begin}–{end}"
        # return f"{begin}-{end}"

In [56]:
df['object_date_combined'] = df.apply(lambda x: combine_date(str(x['Object Begin Date']), str(x['Object End Date'])), axis=1)
df.object_date_combined

Object ID
1              1853
2              1901
3           1909–27
4           1909–27
5           1909–27
            ...    
860869         1904
860870         1904
860871      1768–78
860872    1890–1900
860873    1890–1900
Name: object_date_combined, Length: 477804, dtype: object

In [57]:
# show rows where these differ
df.loc[df['Object Date'] != df['object_date_combined']][['Object Date', 'object_date_combined']]

Object Date object_date_combined
Object ID                                 
35            1893–95               1893–5
36           ca. 1887               1884–7
37           ca. 1785               1782–5
38           ca. 1785               1782–5
44           ca. 1790              1787–90
...               ...                  ...
860867           1758              1753–63
860868           1758              1753–63
860871     after 1773              1768–78
860872           1904            1890–1900
860873           1904            1890–1900

[328702 rows x 2 columns]

from the previous cell output we can see that `Object Date` is kind of random. I am talking about deleting leading digits in the end part, example:
begin = 1755
end = 1758
this is represented in multiple different ways, so like
1755-1758
1755-58
1755-8

there is also a problem with the `ca.`
`ca. 1875` can mean `1872-1875`, but also `1872-1878`

there are also examples like `19th century` which is not that bad, but there are also things like `after 1885`, like what is that supposed to mean, what range after?

## 3B
as I have already talked about in 2B, the number of `|` in all of the artist related columns should be the same, since this is used when there are multiple columns. So if the number of these is different on a row, then it could be an integrity problem, lets check

In [61]:
# find out list of all artist columns
df.columns

Index(['Object Number', 'Is Highlight', 'Is Timeline Work', 'Is Public Domain',
       'Gallery Number', 'Department', 'AccessionYear', 'Object Name', 'Title',
       'Culture', 'Period', 'Dynasty', 'Reign', 'Portfolio', 'Constituent ID',
       'Artist Role', 'Artist Prefix', 'Artist Display Name',
       'Artist Display Bio', 'Artist Suffix', 'Artist Alpha Sort',
       'Artist Nationality', 'Artist Begin Date', 'Artist End Date',
       'Artist Gender', 'Artist ULAN URL', 'Artist Wikidata URL',
       'Object Date', 'Object Begin Date', 'Object End Date', 'Medium',
       'Dimensions', 'Credit Line', 'Geography Type', 'City', 'State',
       'County', 'Country', 'Region', 'Subregion', 'Locale', 'Locus',
       'Excavation', 'River', 'Classification', 'Rights and Reproduction',
       'Link Resource', 'Object Wikidata URL', 'Metadata Date', 'Repository',
       'Tags', 'Tags AAT URL', 'Tags Wikidata URL', 'Object_name_len',
       'object_date_combined'],
      dtype='object')

In [62]:
artist_columns=[
    'Artist Role', 'Artist Prefix', 'Artist Display Name', 'Artist Display Bio', 'Artist Suffix', 'Artist Alpha Sort', 'Artist Nationality', 'Artist Begin Date', 'Artist End Date', 'Artist Gender', 'Artist ULAN URL', 'Artist Wikidata URL'
]

In [64]:
# get number of occurrences for `|` for all artist columns
artist_pipe_counts = {}
for col in artist_columns:
    artist_pipe_counts[col] = df[col].apply(lambda x: str(x).count('|'))

In [74]:
# example
df.loc[artist_pipe_counts['Artist Role'] != artist_pipe_counts['Artist Display Name']]

Empty DataFrame
Columns: [Object Number, Is Highlight, Is Timeline Work, Is Public Domain, Gallery Number, Department, AccessionYear, Object Name, Title, Culture, Period, Dynasty, Reign, Portfolio, Constituent ID, Artist Role, Artist Prefix, Artist Display Name, Artist Display Bio, Artist Suffix, Artist Alpha Sort, Artist Nationality, Artist Begin Date, Artist End Date, Artist Gender, Artist ULAN URL, Artist Wikidata URL, Object Date, Object Begin Date, Object End Date, Medium, Dimensions, Credit Line, Geography Type, City, State, County, Country, Region, Subregion, Locale, Locus, Excavation, River, Classification, Rights and Reproduction, Link Resource, Object Wikidata URL, Metadata Date, Repository, Tags, Tags AAT URL, Tags Wikidata URL, Object_name_len, object_date_combined]
Index: []

[0 rows x 55 columns]

In [79]:
for i in range(0, len(artist_columns)-1):
    a = artist_columns[i]
    b = artist_columns[i+1]
    missmatch_count = len(df.loc[artist_pipe_counts[a] != artist_pipe_counts[b]])
    if missmatch_count > 0:
        print(f"Pipe count does not match for columns `{a}` and `{b}`")


Based on this we know that the count of `|` occurrences is the same for all columns, hence integrity is fine (at least in this case anyway)

# 4. Convert at least **five features** to a proper data type. Choose at least one numeric, one categorical and one datetime. _(1.5 points)_

In [89]:
# lets get all object type columns
df.select_dtypes(include=['object'])

Object Number Gallery Number Object Name  \
Object ID                                            
1            1979.486.1            NaN        Coin   
2            1980.264.5            NaN        Coin   
3              67.265.9            NaN        Coin   
4             67.265.10            NaN        Coin   
5             67.265.11            NaN        Coin   
...                 ...            ...         ...   
860869      23.112.2893            NaN     Drawing   
860870      23.112.2894            NaN     Drawing   
860871      53.600.1434            NaN       Print   
860872      23.112.2895            NaN     Drawing   
860873      23.112.2896            NaN     Drawing   

                                                       Title Culture Period  \
Object ID                                                                     
1                               One-dollar Liberty Head Coin     NaN    NaN   
2                               Ten-dollar Liberty Head Coin     NaN    NaN   
3                                 Two-and-a-Half Dollar Coin     NaN    NaN   
4                                 Two-and-a-Half Dollar Coin     NaN    NaN   
5                                 Two-and-a-Half Dollar Coin     NaN    NaN   
...                                                      ...     ...    ...   
860869                                        Phaeton #24567     NaN    NaN   
860870                Stanhope Phaeton #25538-25539 (#21222)     NaN    NaN   
860871     Forest landscape with cattle drinking, a woman...     NaN    NaN   
860872                              Phaeton with folding top     NaN    NaN   
860873                                      Stanhope Phaeton     NaN    NaN   

          Dynasty Reign Portfolio Constituent ID  ... River Classification  \
Object ID                                         ...                        
1             NaN   NaN       NaN          16429  ...   NaN            NaN   
2             NaN   NaN       NaN            107  ...   NaN            NaN   
3             NaN   NaN       NaN            NaN  ...   NaN            NaN   
4             NaN   NaN       NaN            NaN  ...   NaN            NaN   
5             NaN   NaN       NaN            NaN  ...   NaN            NaN   
...           ...   ...       ...            ...  ...   ...            ...   
860869        NaN   NaN       NaN          16517  ...   NaN       Drawings   
860870        NaN   NaN       NaN          16517  ...   NaN       Drawings   
860871        NaN   NaN       NaN          16489  ...   NaN         Prints   
860872        NaN   NaN       NaN          16517  ...   NaN       Drawings   
860873        NaN   NaN       NaN          16517  ...   NaN       Drawings   

          Rights and Reproduction  \
Object ID                           
1                             NaN   
2                             NaN   
3                             NaN   
4                             NaN   
5                             NaN   
...                           ...   
860869                        NaN   
860870                        NaN   
860871                        NaN   
860872                        NaN   
860873                        NaN   

                                               Link Resource  \
Object ID                                                      
1           http://www.metmuseum.org/art/collection/search/1   
2           http://www.metmuseum.org/art/collection/search/2   
3           http://www.metmuseum.org/art/collection/search/3   
4           http://www.metmuseum.org/art/collection/search/4   
5           http://www.metmuseum.org/art/collection/search/5   
...                                                      ...   
860869     http://www.metmuseum.org/art/collection/search...   
860870     http://www.metmuseum.org/art/collection/search...   
860871     http://www.metmuseum.org/art/collection/search...   
860872     http://www.metmuseum.org/art/collection/search...   
860873     htt

### 4a
as far as datetime, we can choose AssentionYear. It appears to represent years only, but it still doesnt hurt to change to datetime. One of the advantage might be future comparisons

In [84]:
df['AccessionYear']= pd.to_datetime(df['AccessionYear'])

### 4b,c,d
Department seems like a good candidate for categorical data type with 19 unique values`

In [94]:
# sex_category = pd.api.types.CategoricalDtype(categories=['male','female'], ordered=False)
# df['Department'] = df['Department'].astype(sex_category)
df['Department'] = pd.Categorical(df['Department'])

these also look like good candinates
```
Dynasty,403
Reign,389
```

In [93]:
df['Dynasty'] = pd.Categorical(df['Dynasty'])
df['Reign'] = pd.Categorical(df['Reign'])

If I would do the already mentioned cleaning for `Object Name`, then it would also be a good candidate for Categorical data type, but it is definitely not in the current state.

### 4e
as for the numerical data type, I am currently thinking `Constituent ID`, lets try it

In [114]:
try:
    df['Constituent ID']= pd.to_numeric(df['Constituent ID'])
except Exception as e:
    print(e)

Integer out of range. at position 389


Oh, the IDs dont fit into 64bit integer, never mind, lets take a look at smth else.
Maybe `Galery Number`, somebody could consider it as a categorical data type, but I thinking about it more as an ID.

In [103]:
# let's 1st take a look at the data
df.drop_duplicates(subset=['Gallery Number'])

Object Number  Is Highlight  Is Timeline Work  \
Object ID                                                                
1                           1979.486.1         False             False   
33                               64.62         False             False   
35                             04.1a–c          True              True   
42                             60.58.1         False             False   
62                            22.120.1         False             False   
...                                ...           ...               ...   
854924                      2021.326.1         False             False   
855358                    2021.323a, b         False             False   
855895                      48.101.91h         False             False   
857061     NB198.5.J84 J86 1966 Quarto          True             False   
857709                      74.51.6718         False             False   

           Is Public Domain  Gallery Number           Department  \
Object ID                                                          
1                     False             NaN    The American Wing   
33                    False             774    The American Wing   
35                    False             706    The American Wing   
42                     True             724    The American Wing   
62                     True             741    The American Wing   
...                     ...             ...                  ...   
854924                False             253            Asian Art   
855358                False             207            Asian Art   
855895                 True             453          Islamic Art   
857061                False  Watson Library        The Libraries   
857709                 True             171  Greek and Roman Art   

                          AccessionYear                  Object Name  \
Object ID                                                              
1         1970-01-01 00:00:00.000001979                         Coin   
33        1970-01-01 00:00:00.000001964                         Bust   
35        1970-01-01 00:00:00.000001904                         Vase   
42        1970-01-01 00:00:00.000001960                      Andiron   
62        1970-01-01 00:00:00.000001922                      Andiron   
...                                 ...                          ...   
854924    2021-01-01 00:00:00.000000000                       Tangka   
855358    2021-01-01 00:00:00.000000000  Hexagonal vessel with cover   
855895    1948-01-01 00:00:00.000000000           Beads and pendants   
857061                              NaT                          NaN   
857709    1874-01-01 00:00:00.000000000                        Askos   

                                                       Title        Culture  \
Object ID                                                                     
1                               One-dollar Liberty Head Coin            NaN   
33                                   Bust of Abraham Lincoln       American   
35                                            The Adams Vase       American   
42                                                   Andiron       American   
62                                                   Andiron            NaN   
...                                                      ...            ...   
854924                                                   NaN  Central Tibet   
855358     손대현 주칠빙렬무늬 함|孫大鉉　八角牡丹文　器 |Hexagonal vessel wit...          Korea   
855895                                    Beads and Pendants            NaN   
857061                                      66 signs of neon            NaN   
857709          Terracotta askos in the shape of a quadruped        Cypriot   

           ... Rights and Reproduction  \
Object ID  ...                           
1          ...                     NaN   
33         ...                     NaN   
35         ...                     NaN   
42

as far as all of the float values go, all of those have `0` as the decimal value, so those can be safely converted to int.
However, there are some string values as well, since those are not duplicated, lets just replace them with NaN.

In [111]:
# https://stackoverflow.com/questions/53668421/replace-a-string-value-with-nan-in-pandas-data-frame-python
df['Gallery Number']= pd.to_numeric(df['Gallery Number'], errors='coerce')


# 5. Find some outliers and describe your method. _(3 points, depends on creativity)_

### 5.1
I already found a few outliers in 4e. There were 4 rows which had Galery Number as String, which I would consider very outliery

### 5.2


In [115]:
df

Object Number  Is Highlight  Is Timeline Work  Is Public Domain  \
Object ID                                                                   
1            1979.486.1         False             False             False   
2            1980.264.5         False             False             False   
3              67.265.9         False             False             False   
4             67.265.10         False             False             False   
5             67.265.11         False             False             False   
...                 ...           ...               ...               ...   
860869      23.112.2893         False             False              True   
860870      23.112.2894         False             False              True   
860871      53.600.1434         False             False             False   
860872      23.112.2895         False             False              True   
860873      23.112.2896         False             False              True   

           Gallery Number           Department                 AccessionYear  \
Object ID                                                                      
1                     NaN    The American Wing 1970-01-01 00:00:00.000001979   
2                     NaN    The American Wing 1970-01-01 00:00:00.000001980   
3                     NaN    The American Wing 1970-01-01 00:00:00.000001967   
4                     NaN    The American Wing 1970-01-01 00:00:00.000001967   
5                     NaN    The American Wing 1970-01-01 00:00:00.000001967   
...                   ...                  ...                           ...   
860869                NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860870                NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860871                NaN  Drawings and Prints 1953-01-01 00:00:00.000000000   
860872                NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   
860873                NaN  Drawings and Prints 1923-01-01 00:00:00.000000000   

          Object Name                                              Title  \
Object ID                                                                  
1                Coin                       One-dollar Liberty Head Coin   
2                Coin                       Ten-dollar Liberty Head Coin   
3                Coin                         Two-and-a-Half Dollar Coin   
4                Coin                         Two-and-a-Half Dollar Coin   
5                Coin                         Two-and-a-Half Dollar Coin   
...               ...                                                ...   
860869        Drawing                                     Phaeton #24567   
860870        Drawing             Stanhope Phaeton #25538-25539 (#21222)   
860871          Print  Forest landscape with cattle drinking, a woman...   
860872        Drawing                           Phaeton with folding top   
860873        Drawing                                   Stanhope Phaeton   

          Culture  ... Rights and Reproduction  \
Object ID          ...                           
1             NaN  ...                     NaN   
2             NaN  ...                     NaN   
3             NaN  ...                     NaN   
4             NaN  ...                     NaN   
5             NaN  ...                     NaN   
...           ...  ...                     ...   
860869        NaN  ...                     NaN   
860870        NaN  ...                     NaN   
860871        NaN  ...                     NaN   
860872        NaN  ...                     NaN   
860873        NaN  ...                     NaN   

                                               Link Resource  \
Object ID                                                      
1           http://www.metmuseum.org/art/collection/search/1   
2           http://www.metmuseum.org/art/collection/search/2   
3           http://www.metmuseum.org/art/collection/search/3   
4      

In [120]:
sns.boxplot(data=df.Object_name_len)

KeyError: 0

# 6. Detect missing data in at least **three features**, convert them to a proper representation (if they are already not), and impute missing values in at least **one feature**. _(1 + 3 points, depends on creativity)_

# 7. Focus more precisely on the cleaning of the "Medium" feature. As if you were to use it in KNN algorithm later. _(2 points)_

# 8. Focus on the extraction of physical dimensions of each item (width, depth and height in centimeters) from the "Dimensions" feature. _(2 points)_